Detectors/ITSMFT/ITS/tracking/src/VertexerTraits.cxx

```diff
@@ -144,35 +156,74 @@ void VertexerTraits::updateVertexingParameters(const VertexingParameters& vrtPar
+  setNThreads(mVrtParams.nThreads);
 }
 
 void VertexerTraits::computeTracklets()
 {
-  for (int rofId{0}; rofId < mTimeFrame->getNrof(); ++rofId) {
-    trackleterKernelSerial<TrackletMode::Layer0Layer1>(
-      mTimeFrame->getClustersOnLayer(rofId, 0),
-      mTimeFrame->getClustersOnLayer(rofId, 1),
-      mTimeFrame->getIndexTable(rofId, 0).data(),
-      mVrtParams.phiCut,
-      mTimeFrame->getTracklets()[0],
-      mTimeFrame->getNTrackletsCluster(rofId, 0),
-      mIndexTableUtils,
-      rofId,
-      mVrtParams.maxTrackletsPerCluster);
-    trackleterKernelSerial<TrackletMode::Layer1Layer2>(
-      mTimeFrame->getClustersOnLayer(rofId, 2),
-      mTimeFrame->getClustersOnLayer(rofId, 1),
-      mTimeFrame->getIndexTable(rofId, 2).data(),
-      mVrtParams.phiCut,
-      mTimeFrame->getTracklets()[1],
-      mTimeFrame->getNTrackletsCluster(rofId, 1),
-      mIndexTableUtils,
-      rofId,
-      mVrtParams.maxTrackletsPerCluster);
-    mTimeFrame->getNTrackletsROf(rofId, 0) = std::accumulate(mTimeFrame->getNTrackletsCluster(rofId, 0).begin(), mTimeFrame->getNTrackletsCluster(rofId, 0).end(), 0);
-    mTimeFrame->getNTrackletsROf(rofId, 1) = std::accumulate(mTimeFrame->getNTrackletsCluster(rofId, 1).begin(), mTimeFrame->getNTrackletsCluster(rofId, 1).end(), 0);
+
+#pragma omp parallel num_threads(mNThreads)
+  {
+#pragma omp for schedule(dynamic)
+    for (int rofId = 0; rofId < mTimeFrame->getNrof(); ++rofId) {
+      trackleterKernelHost<TrackletMode::Layer0Layer1, true>(
+        mTimeFrame->getClustersOnLayer(rofId, 0),
+        mTimeFrame->getClustersOnLayer(rofId, 1),
+        mTimeFrame->getIndexTable(rofId, 0).data(),
+        mVrtParams.phiCut,
+        mTimeFrame->getTracklets()[0],
+        mTimeFrame->getNTrackletsCluster(rofId, 0),
+        mIndexTableUtils,
+        rofId,
+        0,
+        mVrtParams.maxTrackletsPerCluster);
+      trackleterKernelHost<TrackletMode::Layer1Layer2, true>(
+        mTimeFrame->getClustersOnLayer(rofId, 2),
+        mTimeFrame->getClustersOnLayer(rofId, 1),
+        mTimeFrame->getIndexTable(rofId, 2).data(),
+        mVrtParams.phiCut,
+        mTimeFrame->getTracklets()[1],
+        mTimeFrame->getNTrackletsCluster(rofId, 1),
+        mIndexTableUtils,
+        rofId,
+        0,
+        mVrtParams.maxTrackletsPerCluster);
+      mTimeFrame->getNTrackletsROf(rofId, 0) = std::accumulate(mTimeFrame->getNTrackletsCluster(rofId, 0).begin(), mTimeFrame->getNTrackletsCluster(rofId, 0).end(), 0);
+      mTimeFrame->getNTrackletsROf(rofId, 1) = std::accumulate(mTimeFrame->getNTrackletsCluster(rofId, 1).begin(), mTimeFrame->getNTrackletsCluster(rofId, 1).end(), 0);
+    }
+#pragma omp single
+    mTimeFrame->computeTrackletsScans(mNThreads);
+#pragma omp single
+    mTimeFrame->getTracklets()[0].resize(mTimeFrame->getTotalTrackletsTF(0));
+#pragma omp single
+    mTimeFrame->getTracklets()[1].resize(mTimeFrame->getTotalTrackletsTF(1));
```

Comment:
These lines are outside of the MT-ed block, why do you need `omp single`?

Reply:
My understanding was that when you enter an omp parallel section it spanws the threads and keep them alive until it ends then tries to assign the instruction included in the block to available threads according to scheduling.

Reply:
But your parallel block is the `for` loop only, afterwards OPM does nothing

Reply:
but I have two of them

Reply:
Not sure what you mean. https://github.com/AliceO2Group/AliceO2/blob/ed25a1e47c5fd105013061949b6fc2582396ab1c/Detectors/ITSMFT/ITS/tracking/src/VertexerTraits.cxx#L165 just reserves N threads which are applied only to https://github.com/AliceO2Group/AliceO2/blob/ed25a1e47c5fd105013061949b6fc2582396ab1c/Detectors/ITSMFT/ITS/tracking/src/VertexerTraits.cxx#L167-L193 block then, independently, to https://github.com/AliceO2Group/AliceO2/blob/ed25a1e47c5fd105013061949b6fc2582396ab1c/Detectors/ITSMFT/ITS/tracking/src/VertexerTraits.cxx#L201-L226
The linear code outside of the `for` loops is not affected by the OMP, so no `#pragma omp single` is needed

Reply:
> The linear code outside of the `for` loops is not affected by the OMP, so no `#pragma omp single` is needed

Ok, I might be wrong, my understanding was that when I am inside a `#pragma omp parallel { .... }` block (quoting [this](https://www.ibm.com/docs/en/zos/2.4.0?topic=processing-pragma-omp-parallel) reference):

```
When a parallel region is encountered, a logical team of threads is formed. 
Each thread in the team executes all statements within a parallel region except for work-sharing constructs. 
Work within work-sharing constructs is distributed among the threads in a team.
```
Since we are still in the `omp parallel` block, I'd expect to have the threads to be alive and doing the same things.
This is why I explicitly constrained a single thread to do that (I had crashes due to racing conditions).

Reply:
@shahor02 @mconcas 
I think Matteo understood correctly, since you have 2 parallel blocks:
```
#pragma omp parallel // from here on we run with multiple threads
{
#pragma omp for
  for (int i = 0; i < 10; i++) {
    [...] // This for loop runs with the threads declared in the outer block
  }
#pragma omp single
  {
    foo(); // These functions are only called by one of the threads in the outer block
    bar();
  }
#pragma omp for
  for (int i = 0; i < 12; i++) {
    [...] // Again running multi-threaded, using the same threads as above.
  }
}
```

The diference to:
```
#pragma omp parallel for
for (int i = 0; i < 10; i++) {
  [...]
}
foo();
bar();
#pragma omp parallel for
for (int i = 0; i < 12; i++) {
  [...]
}
```
is that in the first case, the threads are not synchronized up until the end of the parallel block.
In the second case, threads are synchronized when exiting the first parallel block, and before the main thread calls `foo()`

Note the difference between `#pragma omp for` and `#pragma omp parallel for`. The first uses threads of an existing parallel block, while the later would spawn new threads (or take existing threads from a thread pool).
E.g.
```
#pragma omp parallel num_threads(4)
{
#pragma omp parallel for num_threads(4)
  for (int i = 0; i < 10; i++) {
    [...]
  }
}
```
Would use 16 threads to run the full loop 4 times, each time 4-fold multi-threaded.

Reply:
OK, then I withdraw my comment. Actually, from https://stackoverflow.com/questions/1448318/omp-parallel-vs-omp-parallel-for I understand that w/o `#pragma omp single` all threads (not only the main one) will execute the the code outside of the `for` loops but inside the `omp parallel` block

Reply:
yes, exactly